### PR TITLE
Add DBR 17.3 Delta CTAS/RTAS support and fix optimized writes [databricks]

### LINKED ISSUE #15263: [FEA] Delta Lake DB-17.3: Enable GPU data-file writes for managed CTAS/RTAS
**Is your feature request related to a problem? Please describe.**

The existing DBR 17.3 Delta write support does not fully accelerate managed atomic CTAS/RTAS operations. In particular, the following customer pattern can retain a CPU Delta data-file writer:

```
df.write \
  .format("delta") \
  .mode("overwrite") \
  .option("overwriteSchema", "true") \
  .saveAsTable(target)
```

DBR 17.3 represents these operations using outer catalog commands such as:
`AtomicCreateTableAsSelectExec`
`AtomicReplaceTableAsSelectExec`

The expensive data-writing work is performed by a nested DBR WriteIntoDeltaCommand / WriteFilesExec. Accelerating only the input query or changing the outer plan label, is insufficient if the nested Delta file writer still uses the CPU/Parquet-MR path.

### PR DESCRIPTION
**Pull request by nartal1**
Fixes #15263

### Description

The initial Delta Lake write support for Databricks Runtime 17.3 kept atomic CTAS and RTAS operations on CPU. In addition, DBR 17.3 optimized writes could lose their target-partitioning marker during CPU-to-GPU shuffle conversion, preventing the optimized-write exchange from being reconstructed and resized correctly on GPU.

This PR enables the supported DBR 17.3 Delta CTAS/RTAS paths on GPU and fixes the optimized-write shuffle conversion.

In this PR:

- Converts supported `AtomicCreateTableAsSelectExec` and `AtomicReplaceTableAsSelectExec` plans to the corresponding GPU atomic wrappers.
- Retains DBR’s native catalog, staged-table, commit, abort, cache invalidation, and table-replacement semantics.
- Keeps the query and Delta data-file write path GPU-eligible inside the atomic operation.
- Preserves `DeltaOptimizedWritePartitioning` as the target partitioning while advertising the derived physical partitioning as the shuffle output.
- Preserves optimized-write marker-local configuration overrides during CPU-to-GPU conversion.
- Keeps the target and physical partitioning contracts consistent through tree copies, explicit partition-count changes, and AQE reconstruction.
- Adds structural plan validation based on the `DELTA_OPTIMIZED_WRITE` shuffle origin, preventing unrelated GPU exchanges from satisfying the optimized-write plan check.
- Adds focused conversion and AQE coverage for the DBR 17.3 optimized-write shuffle.

### Expected fallback behavior

The implementation continues to fall back to CPU for table-creation features whose DBR metadata or catalog semantics have not been qualified for the GPU create path, including:

- Row filters and column masks.
- Liquid clustering and auto-TTL.
- Catalog-owned tables and coordinated commits.
- Explicit or session-default deletion-vector creation settings that require unsupported metadata handling.

When Delta optimized writes are enabled with AQE, the `DELTA_OPTIMIZED_WRITE` exchange can run on GPU. With optimized writes enabled and AQE disabled, DBR uses its non-AQE `DeltaOptimizedWriterExec`; that path remains an expected CPU fallback.

### Performance

The scaled workload used:

- Databricks Runtime 17.3, Spark 4.0, and Scala 2.13.
- Eight `g4dn.4xlarge` T4 workers and a matching driver.
- A deterministic SF100-class Delta source with 452,433,852 rows across 12 source tables and 27.36 GB of compressed source data.
- Fourteen representative model-build transformations producing 425,795,424 final rows.

This is an SF100-class workload and is not an official TPC-DS SF100 benchmark.

#### GPU Delta writer benefit

Delta optimized writes were disabled in this comparison so that it isolates the Delta writer change.

| Configuration | Delta writer | Fresh CTAS | Replacement RTAS |
| --- | --- | ---: | ---: |
| RAPIDS 26.06 release | CPU fallback | 138.819 s | 133.876 s |
| 26.08 SNAPSHOT CPU-writer control | CPU | 128.665 s | 126.982 s |
| 26.08 SNAPSHOT with this GPU write path | GPU | 92.874 s | 90.778 s |

Compared with the same-26.08 CPU-writer control, GPU Delta writing was:

- 1.385x faster for fresh table creation.
- 1.399x faster for replacement.

Compared with the RAPIDS 26.06 released behavior, it was:

- 1.495x faster for fresh table creation, a 33.1% wall-time reduction.
- 1.475x faster for replacement, a 32.2% wall-time reduction.

Physical-plan validation confirmed that the 26.06 run accelerated the query operators but crossed through `GpuColumnarToRow` into CPU `WriteIntoDeltaCommand`/`WriteFiles`. The 26.08 GPU result used the GPU atomic wrapper and GPU Delta writer.

#### Delta optimized-write tradeoff

spark.databricks.delta.optimizeWrite.enabled=false/true using the build  jar from this PR. 

| Delta optimized writes | Fresh median | Replacement median | Output files |
| --- | ---: | ---: | ---: |
| Off | 95.724 s | 95.318 s | 2,004 |
| On, with AQE | 117.304 s | 116.041 s | 84 |

Enabling optimized writes increased write time by 22.5% for fresh creation and 21.7% for replacement. In exchange, it reduced the output from 2,004 files to 84 files, a 23.86x reduction.

This was not caused by CPU fallback: the optimize-on runs used GPU atomic wrappers, GPU Delta writing, and GPU `DELTA_OPTIMIZED_WRITE` exchanges. The additional time represents the actual repartitioning and file-sizing work required to produce the more compact layout.

### Checklists

Documentation
- [ ] Updated for new or modified user-facing features or behaviors
- [x] No user-facing change

Testing
- [x] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
      (Please provide the names of the existing tests in the PR description.)
- [ ] Not required

Performance
- [x] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [ ] Not required

### CODE REVIEW

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
build

File: integration_tests/src/main/python/delta_lake_write_test.py [hunk 1]
```diff
@@ -212,6 +212,280 @@ def test_delta_write_round_trip_managed(spark_tmp_table_factory, enable_deletion
...
+            .write.format("delta").mode("overwrite").saveAsTable(source_table),
+        conf=conf)
+
+    def execute_atomic_sql(spark, table, replace):
+        effective_optimize_write = spark.conf.get(
+            "spark.databricks.delta.optimizeWrite.enabled")
+        assert effective_optimize_write.lower() == str(optimized_write).lower()
+        effective_aqe = spark.conf.get("spark.sql.adaptive.enabled")
+        assert effective_aqe.lower() == str(aqe_enabled).lower()
+        command = "CREATE OR REPLACE TABLE" if replace else "CREATE TABLE"
```
Comment (firestarman) [inline integration_tests/src/main/python/delta_lake_write_test.py:541]: nit: Could we add coverage for the existing-table branches of the new atomic CTAS/RTAS wrappers? In particular, `CREATE TABLE IF NOT EXISTS`, the existing-table error for plain `CREATE`, and `REPLACE TABLE` without `OR CREATE`.

File: integration_tests/src/main/python/delta_lake_write_test.py [hunk 2]
```diff
@@ -212,6 +212,280 @@ def test_delta_write_round_trip_managed(spark_tmp_table_factory, enable_deletion
...
+        "spark.sql.execution.sortBeforeRepartition": "true",
+    })
+    cpu_ctas_table = spark_tmp_table_factory.get() + '_sql_ctas_cpu'
+    gpu_ctas_table = spark_tmp_table_factory.get() + '_sql_ctas_gpu'
+    cpu_rtas_table = spark_tmp_table_factory.get() + '_sql_rtas_cpu'
+    gpu_rtas_table = spark_tmp_table_factory.get() + '_sql_rtas_gpu'
+    source_table = spark_tmp_table_factory.get() + '_sql_source'
+
+    with_cpu_session(
+        lambda spark: spark.range(81920).selectExpr(
```
Comment (firestarman) [inline integration_tests/src/main/python/delta_lake_write_test.py:528]: nit: Could we add a focused zero-row CTAS/RTAS case? The new tests only exercise non-empty inputs, so the schema-only/empty commit path is not covered.

File: sql-plugin/src/main/spark400db173/scala/org/apache/spark/sql/execution/datasources/v2/rapids/GpuAtomicReplaceTableAsSelectExec.scala [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  StagingTableCatalog, Table, TableCatalog, TableInfo, TableWritePrivilege}
+import org.apache.spark.sql.connector.expressions.Transform
+import org.apache.spark.sql.errors.QueryCompilationErrors
+import org.apache.spark.sql.execution.datasources.v2.V2CreateTableAsSelectBaseExec
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * GPU wrapper for DBR 17.3 atomic RTAS.
+ *
+ * DBR's native catalog remains responsible for stage/commit/abort and therefore for Unity
```
Comment (firestarman) [inline sql-plugin/src/main/spark400db173/scala/org/apache/spark/sql/execution/datasources/v2/rapids/GpuAtomicReplaceTableAsSelectExec.scala]: nit: Could we clarify that only the native stage/commit/abort mechanics are retained here, while catalog-owned and coordinated-commit table creation remains blocked pending GPU qualification? The current wording can read as though those semantics are supported.

File: sql-plugin/src/main/spark400db173/scala/org/apache/spark/sql/execution/datasources/v2/rapids/GpuAtomicCreateTableAsSelectExec.scala [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  override def supportsColumnar: Boolean = false
+
+  private def loadForInsert(): Table = {
+    catalog.loadTable(ident, Set(TableWritePrivilege.INSERT).asJava)
+  }
+
+  override protected def run(): Seq[InternalRow] = {
+    if (catalog.tableExists(ident)) {
+      if (ifNotExists) {
+        return Nil
```
Comment (firestarman) [inline sql-plugin/src/main/spark400db173/scala/org/apache/spark/sql/execution/datasources/v2/rapids/GpuAtomicCreateTableAsSelectExec.scala]: nit: Could this avoid explicit `return`s and use expression-style branching or matching instead? The repository’s Scala convention prefers the final expression as the result.

File: sql-plugin/src/main/spark400db173/scala/org/apache/spark/sql/execution/datasources/v2/rapids/GpuAtomicDeltaWriteContext.scala [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+package org.apache.spark.sql.execution.datasources.v2.rapids
+
+import org.apache.spark.SparkContext
+
+/** Limits DBR's nested Delta data-writing command to a validated atomic CTAS/RTAS call stack. */
+object GpuAtomicDeltaWriteContext {
+  private val activeKey = "spark.rapids.sql.delta.atomicWrite.active"
+
+  def isActive: Boolean = SparkContext.getActive
+    .exists(_.getLocalProperty(activeKey) == "true")
```
Comment (firestarman) [inline sql-plugin/src/main/spark400db173/scala/org/apache/spark/sql/execution/datasources/v2/rapids/GpuAtomicDeltaWriteContext.scala]: nit: Could this context use a private per-JVM token rather than the public literal `"true"`? Application code can set Spark local properties, so a fixed value makes the qualification gate externally reproducible.

**Review by firestarman — COMMENTED** [review]
lgtm

**Review by firestarman — DISMISSED** [review]

**Comment by nvauto** [comment]
**NOTE: release/26.08 has been created from main. Please retarget your PR to release/26.08 if it should be included in the release.**

**Comment by nartal1** [comment]
build

**Review by firestarman — APPROVED** [review]
LGTM
